engine/src/test/java/io/camunda/zeebe/process/test/engine/EngineStateMonitorTest.java

```diff
@@ -161,7 +165,7 @@ public void close() {}
 
     @Override
     public boolean hasNext() {
-      return position < lastEventPosition;
+      return !stateLocked && position < lastEventPosition;
```

Comment:
:x: I think this is a bit misleading in terms of naming and what it does. It seems to lock only one of the states. 

I think it would be clearer to either rename `stateLocked` or to capture the state at the moment it is locked:
````
setStateLocked(stateLocked) {
  this.stateLocked = stateLocked;
  if (stateLocked) {
     capturedState = position < lastPosition; 
  }
}

hasNext()  {
 return stateLocked ? capturedState : position < lastPosition
}
````

Reply:
I'm not sure what you mean here. `stateLocked` definitely locks both of the states.

If the state is locked `!stateLocked && position < lastEventPosition;` will always resolve as `false`. This has as a result that the `EngineStateMonitor` will never update it's position. This means that if the positions are equal, the state is locked to idle. If the positions are not equal the state is locked to busy.
```java
synchronized (reader) {
      while (reader.hasNext()) { // <--- this will resolve to false so the lastEventPosition is never updated.
        lastEventPosition = reader.next().getPosition();
      }
    }
```

Reply:
Hmm, I see what you mean. The thing that is confusing me is that `hasNext` is always false when `stateLocked=true`. Which seems wrong, in my mind. And the only thing it does work is indirectly because of taking into account implementation details of how this method is used as a guard around the call to next.

Maybe another way could be to implement it like this:
```
    @Override
    public LoggedEvent next() {
     if !stateLocked {
        position++;
    }
      final LoggedEvent event = mock(LoggedEvent.class);
      when(event.getPosition()).thenReturn(position);
      return event;
    }
```

So then `hasNext` would give always the correct result, but one couldn't really advance. But maybe this then results in endless loops 🤔

Reply:
Yes, this solution will result in infinite loops.

I don't like the other solution either though:
```java
setStateLocked(stateLocked) {
  this.stateLocked = stateLocked;
  if (stateLocked) {
     capturedState = position < lastPosition; 
  }
}

hasNext()  {
 return stateLocked ? capturedState : position < lastPosition
}
```
This way it seems like. the state is being managed by the reader which is not the case.

The way it is currently makes sense in my mind. The `EngineStateMonitor` asks if there is any more events it should process. Because the state is locked we override this by saying, no there isn't anything. Thus, keeping the state of the `EngineStateMonitor` the same.

Reply:
What about we time box this, each of us thinks 30mins hard about the problem. If we find no better solution, we leave it as it is?

Reply:
Sounds like a plan

Reply:
I think this is goo enough for now